apputils/eval.go

```diff
@@ -0,0 +1,292 @@
+	GreaterThan EvalOperator = "gt"
+	// GreaterThanOrEqual is a typed representation of the >= operator.
+	GreaterThanOrEqual EvalOperator = "gte"
+	// Equal is a typed representation of the == operator.
+	Equal EvalOperator = "eq"
+	// NotEqual is a typed representation of the != operator.
+	NotEqual EvalOperator = "neq"
+	// Regex is a typed representation of the regular expression operator.
+	Regex EvalOperator = "regex"
+)
+
+// ToReadableString returns a string representation of an evaluable operator.
+func (operator EvalOperator) ToReadableString() string {
+	switch operator {
+	case LessThan:
+		return "less than"
+	case LessThanOrEqual:
+		return "less than or equal to"
+	case GreaterThan:
+		return "greater than"
+	case GreaterThanOrEqual:
+		return "greater than or equal to"
+	case Equal:
+		return "equal to"
+	case NotEqual:
+		return "not equal to"
+	case Regex:
+		return "a match on the regular expression"
+	}
+	return ""
+}
+
+// Evaluate returns whether 'readingValue' is [insert operator here] 'conditionValue'.
+func (operator EvalOperator) Evaluate(readingValue, conditionValue any) (bool, error) {
+	if IsMap(readingValue) && IsMap(conditionValue) {
+		mapReadingValue, ok := readingValue.(bson.M)
+		if !ok {
+			return false, errors.New("cast assertion failed: reading must be bson.M type")
+		}
+		mapConditionValue, ok := conditionValue.(map[string]interface{})
+		if !ok {
+			return false, errors.New("cast assertion failed: condition must be map[string]interface{} type")
+		}
+		return operator.evaluateMaps(mapReadingValue, mapConditionValue)
+	}
+	if IsNumber(readingValue) && IsNumber(conditionValue) {
+		readingValueFloat, err := ConvertToFloat64(readingValue)
+		if err != nil {
+			return false, err
+		}
+		conditionValueFloat, err := ConvertToFloat64(conditionValue)
+		if err != nil {
+			return false, err
+		}
+		return operator.evaluateFloats(readingValueFloat, conditionValueFloat)
+	}
+	if IsSlice(readingValue) && IsSlice(conditionValue) {
+		// Compare slices of numerical values (so far is the only needed use case, but extend here if
+		// support for non-numerical slice comparing is needed).
+		readingValueSlice, err := ConvertToFloat64Slice(readingValue)
+		if err != nil {
+			return false, err
+		}
+		conditionValueSlice, err := ConvertToFloat64Slice(conditionValue)
+		if err != nil {
+			return false, err
+		}
+		return evaluateLists[float64](operator, readingValueSlice, conditionValueSlice)
+	}
+	return operator.evaluateValues(readingValue, conditionValue)
+}
+
+// IsMap returns whether value is a map type.
+func IsMap(value interface{}) bool {
+	return value != nil && (reflect.TypeOf(value).Kind() == reflect.Map)
+}
+
+// IsSlice returns whether value is a list of any type.
+func IsSlice(value interface{}) bool {
+	return value != nil && (reflect.TypeOf(value).Kind() == reflect.Slice)
+}
+
+// IsNumber takes any value and returns whether its a numeric type.
+func IsNumber(value interface{}) bool {
```

Comment:
I wish go had better ergonomics around stuff like this (offered this method and the two above natively) 😮‍💨 .